Meeting scheduling functionality

jaimehisao opened the pull request:
There is a lot of debug output. This will be removed in the future.

jaimehisao committed:
Changed FindMeetingQuery

jaimehisao committed:
Adjusting TimeRanges where meetings dont overlap works

jaimehisao commented:
Initial draft of the PR. Comments are a work in progress and also the algorithm isn't optimized yet. Focusing on having a working solution and then optimize the code.

ssmall reviewed (commented):
Just leaving a comment to set review state -- re-request review when it's ready with the latest changes

jaimehisao committed:
Before changing full structure

jaimehisao committed:
Changed to array implementation

jaimehisao committed:
Removed uneccesary code

jaimehisao committed:
Cleaned code

jaimehisao commented:
Changed from having a "split" function to having an array with every minute of the day.

anniefu commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -15,9 +15,80 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
```
[nit] duplicate import

jaimehisao replied:
actually, when I remove either of those I get a compilation error...

ssmall replied:
one is `Collection` and the other is `Collections` ... not a duplicate, I think

anniefu replied:
Oh, good catch. Weird that Java has both haha.

anniefu commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -15,9 +15,80 @@
+   * Returns {@code TimeRange} when a meeting can happen based on other events and the
+   * meeting participants involved.
+   * @param events {@code Collection} of events happening.
+   * @param request {@code MeetingRequest} for a meeting
```
Nice use of Java doc features! Could you add the big-O runtime of your algorithm too?

anniefu commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -15,9 +15,80 @@
+    ArrayList<Event> localEvents = new ArrayList<Event>(events);
+    boolean[] availableTimes = new boolean[MINUTES_IN_A_DAY+1];
+
+    //Meeting is longer than a day, that can't happen
```
[nit] here and in the rest of this file, leave a space between comments "//" and start of comment: https://google.github.io/styleguide/javaguide.html#s4.8.6-comments

ssmall reviewed (changes requested):
This PR doesn't appear to include any changes to unit tests -- please include the additional unit tests described in https://github.com/jaimehisao/step/blob/master/walkthroughs/week-5-tdd/project/calendar-walkthrough.md#writing-more-tests

jaimehisao commented:
Yes, I'm working on those right now.

jaimehisao committed:
Added Optional Atendee Full Day Event Test

jaimehisao committed:
Added morning meeting optional attendee test

anniefu commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -15,9 +15,80 @@
+      return possibleTimes;
+    }
+
+    //No atendees on the request
```
[nit] *attendees, and on L59.

anniefu commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -15,9 +15,80 @@
+      if(!Collections.disjoint(event.getAttendees(), request.getAttendees())){
+        Arrays.fill(availableTimes, event.getWhen().start(), event.getWhen().end(), true);
+      }else{
+        localEvents.remove(event);
```
A couple of performance notes here, a call to remove an object from an ArrayList is going to cause an `O(n)` search for the object and an `O(n)` removal for random deletion from an array.

For best performance on random deletion, you probably want a LinkedList which gives `O(1)` removal.

jaimehisao replied:
removed!

anniefu commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -15,9 +15,80 @@
+    //Find free time in the array and if they fit the meeting, create the time range
+    int intervalDuration = 0;
+    for(int i = 0; i < availableTimes.length; i++){
+      if(!availableTimes[i]){
```
I'm not sure I understand the goal here.

Because of the `!availableTimes[i]` check, it seems like `intervalDuration` is supposed to represent a period of **unavailable** times, but then these are added to the `possibleTimes` array?

jaimehisao replied:
availableTimes holds true where there's a meeting. Wrong choice of name, which has been fixed.

anniefu commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -15,9 +15,80 @@
+    for(Event event : events){
+      //If event has no atendees involved in the request.
+      if(!Collections.disjoint(event.getAttendees(), request.getAttendees())){
+        Arrays.fill(availableTimes, event.getWhen().start(), event.getWhen().end(), true);
```
How do TimeRanges where there are no `events` at all get marked as `true`? It seems like if there's no `event` during a TimeRange, it should be marked as available by default?

anniefu commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -15,9 +15,80 @@
+    }
+
+    //If there is no event, then the whole day is free.
+    if(localEvents.size() == 0){
```
I think you can actually omit this check and removing from localEvents, since you can actually detect this scenario while iterating over `availableTimes`. Ideally, in the case where there are no Events where eventAttendees and mandatoryAttendees overlap, the result of `possibleTimes` after iterating over `availableTimes` is one TimeRange equivalent to WHOLE_DAY.

jaimehisao replied:
Yes! Removing from localEvents was a temporary fix, especially for the performance issues. Neither the check, nor the removal of items exist in the newest version.

anniefu commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -15,9 +15,80 @@
+    final int MINUTES_IN_A_DAY = 1440;
+    ArrayList<TimeRange> possibleTimes = new ArrayList<TimeRange>();
+    ArrayList<Event> localEvents = new ArrayList<Event>(events);
+    boolean[] availableTimes = new boolean[MINUTES_IN_A_DAY+1];
```
Does `availableTimes` actually end up being an array where all available times are marked with `true`?

I'm not certain that's true and I think it's a core piece of understanding what the algorithm is doing.

jaimehisao replied:
I was actually marking the busy times. It was available times at first but then I marked the times with meetings.
Will rename accordingly.

anniefu reviewed (commented):
Good use of comments!

jaimehisao committed:
Changed to take optional attendees into consideration

jaimehisao commented:
Changed to now take into consideration optional attendees and added the first 3 of the required extra tests.
Also, per @anniefu 's comments, renamed the availableTimes array because it represents the times that are taken up by meetings, instead of the times available, my bad.

jaimehisao committed:
Added Third Test

jaimehisao committed:
Added No Mandatory Only Optionals Test

jaimehisao committed:
Added final required test

jaimehisao committed:
Added Comments

jaimehisao committed:
Removed localEvents since it added unnecessary space usage

jaimehisao committed:
Removed unnecessary variable

jaimehisao committed:
Added more comments & runtime

jaimehisao committed:
Added more comments & runtime

ssmall commented on walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:
```diff
@@ -270,5 +274,148 @@ public void notEnoughRoom() {
+    // Have one mandatory person, but make it so that there is just enough room at one point in the day to
+    // have the meeting. With an additional optional attendee that will be ignored.
+    //
+    // Optional  :     |-C-|
```
Nit: I believe this should be `B`, not `C`

ssmall commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -14,10 +14,108 @@
+    ArrayList<TimeRange> possibleTimes = new ArrayList<TimeRange>();
+    // Find free time in the array and if they fit the meeting, create the time range
+    int availabilityInterval = 0;
+    for(int minuteOfTheDay = 0; minuteOfTheDay < MINUTES_IN_A_DAY+1; minuteOfTheDay++){
```
nit: instead of
```java
minuteOfTheDay < MINUTES_IN_A_DAY+1
```
you could just write 
```java
minuteOfTheDay <= MINUTES_IN_A_DAY
```

anniefu replied:
Since you're actually iterating over busyTimes, maybe just `minuteOfTheDay < len(busyTimes)` (or whatever the length function is in Java).

It's possible that someone in the future changes busyTimes to not be MINUTES_IN_A_DAY long.

ssmall commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -14,10 +14,108 @@
+  /**
+   * Returns {@code TimeRange} when a meeting can happen based on other events and the
+   * meeting participants involved.
+   * Runtime: O(N*M) -> The runtime of {@code query} will be proportional to the amount of 
+   * events (N) and attendees (M). This is seen when calling {@code Collections.disjoint} because it 
+   * is called once for the mandatory attendees and once for the optional attendees and
+   * that happens for every event.
```
Great analysis!

ssmall commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -14,10 +14,108 @@
+    for(Event event : events){
+      // If a mandatory attendee is busy, the meeting can't happen, so time is blocked out on both arrays.
+      if(!Collections.disjoint(event.getAttendees(), request.getAttendees())){
+        Arrays.fill(busyTimesMandatoryAttendees, event.getWhen().start(), event.getWhen().end(), true); 
+        Arrays.fill(busyTimesOptionalAttendees, event.getWhen().start(), event.getWhen().end(), true); 
+      }else if(!Collections.disjoint(event.getAttendees(), request.getOptionalAttendees())){ 
+        // If an optional attendee is busy, only the optional attendee schedule is blocked.
+        Arrays.fill(busyTimesOptionalAttendees, event.getWhen().start(), event.getWhen().end(), true); 
```
Good use of collection utility functions!

ssmall reviewed (approved):
Looks great to me! Would be good to get @anniefu's approval as well before merging.

anniefu commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -14,10 +14,108 @@
+      if(!Collections.disjoint(event.getAttendees(), request.getAttendees())){
+        Arrays.fill(busyTimesMandatoryAttendees, event.getWhen().start(), event.getWhen().end(), true); 
+        Arrays.fill(busyTimesOptionalAttendees, event.getWhen().start(), event.getWhen().end(), true); 
+      }else if(!Collections.disjoint(event.getAttendees(), request.getOptionalAttendees())){ 
```
[nit] leave spaces on around "else if": https://google.github.io/styleguide/javaguide.html#s4.1.2-blocks-k-r-style

anniefu commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -14,10 +14,108 @@
+    int availabilityInterval = 0;
+    for(int minuteOfTheDay = 0; minuteOfTheDay < MINUTES_IN_A_DAY+1; minuteOfTheDay++){
+      // If the minute is not busy, add a free minute to the counter.
+      if(!busyTimes[minuteOfTheDay]){
```
[nit] here and elsewhere, leave a space between "if" and open parenthesis.

anniefu commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -14,10 +14,108 @@
+    ArrayList<TimeRange> possibleTimes = new ArrayList<TimeRange>();
+    // Find free time in the array and if they fit the meeting, create the time range
+    int availabilityInterval = 0;
+    for(int minuteOfTheDay = 0; minuteOfTheDay < MINUTES_IN_A_DAY+1; minuteOfTheDay++){
```
[nit] also spaces between "for" and start parenthesis.

anniefu commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -14,10 +14,108 @@
+    }
+
+    // In case there is no event later, this will add the time since the last event until EoD.
+    if(busyTimes.length - availabilityInterval  >= duration && availabilityInterval != 1){
```
Wouldn't availabilityInterval == 1 if the last minute of the day is available? Shouldn't that be a possibleTime if my requested meeting duration is only 1 minute?

jaimehisao replied:
Yeah but the array is created with an extra "slot" so technically that 1 is accounting for that. When that `availabilityInterval == 1` isn't there a TimeRange of [1440,1440) is created. So it's a meeting that has a duration of 0

jaimehisao replied:
I went ahead and added a test for this case. It contains a meeting that lasts from START_OF_DAY all the way to END_OF_DAY (non inclusive of 11:59PM), and the test checks if query returns a single meeting from 1439 to 1440, which is the EoD

anniefu replied:
Cool, sounds good. Thanks for explaining and adding the test case!

anniefu reviewed (changes requested):
Nice! Makes a lot more sense now that availableTimes is busyTimes :) 

My approval is just waiting on the one question about when `availabilityInterval == 1`.

jaimehisao committed:
nit fixes

jaimehisao committed:
Added test for EoD Meeting w/1Min Duration
